### PR TITLE
[Bugfix][Structured Outputs] Reject empty JSON schemas

### LINKED ISSUE #52011: [bug]: empty JSON schema `{}` causes unbounded number death-loop garbled output
# Empty JSON schema `{}` causes unbounded number death-loop garbled output

## Summary

When a structured-output request uses an empty JSON schema `{"json": {}}` (equivalent to "accept any JSON value"), vLLM compiles it into a grammar whose `number` branch is unbounded and cannot be terminated. The model falls into a number death-loop, emitting up to `max_tokens` of pure digits (a single giant valid JSON number), instead of a normal JSON value. `reasoning_content` (thinking phase) is unaffected — only the final `content` is garbled.

This reproduces on **GPU** as well as Ascend NPU (confirmed by a colleague on GPU vLLM; sglang does **not** reproduce). So it is a vLLM upstream issue, not backend/platform-specific.

## Environment

- vLLM version: `v0.20.3.dev47+g8d52ba3cd` (branch `hpu-release/release-v0.20.2-20260601`, commit `8d52ba3cd`)
- Model: GLM-5.1 (w8a8), with `--reasoning-parser glm45` (thinking enabled)
- Backend: structured_outputs `auto` (resolves to `guidance`/`llguidance`; also reproduces with `xgrammar`/`outlines`)
- Hardware: Ascend HPU 910A3 (also reproduced on GPU)
- Deployment: prefill-decode disaggregation (pd-sep) — but the bug is not pd-sep-specific; it is in the structured-output schema compilation path

### Request

```bash
curl -s http://<endpoint>/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "auto",
    "messages": [{"role": "user", "content": "你好"}],
    "structured_outputs": {"json": {}},
    "stream": false,
    "max_tokens": 51200
  }'
```

The key trigger is `structured_outputs: {"json": {}}` — an **empty JSON schema dict**. In JSON Schema semantics, `{}` (no `type` keyword) means "accept any JSON value" (object/array/string/number/boolean/null).

### Actual (buggy) result

```json
{
  "choices": [{
    "message": {
      "role": "assistant",
      "content": "1.1478652e-0852808689391498581278959278115265275173994480839003085814898587224337225458316874850289548247142149196596559984359243185529485706...",
      "reasoning_content": "收到用户发来的\"你好\"这样简单的问候，需要礼貌地回应并引导对话..."
    },
    "finish_reason": "length"
  }],
  "usage": {"completion_tokens": 51200, ...}
}
```

- `content` is a **single giant JSON number** (`-1751849...e+2011010...` in the original badcase, 118496 chars, all valid JSON number characters — only `0-9` plus one each of `-`, `.`, `e`, `+`; zero illegal JSON-number characters).
- `reasoning_content` is **completely normal** (fluent Chinese).
- `finish_reason`: `length` (ran to `max_tokens`), `completion_tokens`: 51200 (full budget consumed).
- With `max_tokens=200`, `content` is `1.1478652e-085280868939...` (200 tokens of digits), `finish_reason=length`.

### Expected result

Either:
- Reject the empty schema with a 400 error (like vLLM already does for `{"json": ""}` empty string), or
- Normalize `{}` to `{"type": "object"}` (json_object semantics, which has a `}` terminator and produces a normal `{}`).

sglang handles this case (does not produce the death-loop), suggesting vLLM is missing an empty-schema rejection/normalization step that sglang has.

## Root cause (code-level analysis)

Three compounding defects, all in `vllm/` (upstream, not vllm-ascend):

### 1. No empty-schema rejection/normalization anywhere

- `StructuredOutputsParams.__post_init__` (`vllm/sampling_params.py`, class at line 40, `__post_init__` at line 58) only validates mutual-exclusivity and "at least one constraint". An empty dict `{}` as the `json` field passes (`self.json is not None` → count=1 → valid). No special handling.
- `_validate_structured_outputs` (`vllm/sampling_params.py`, ~line 831-975) has early-reject guards for empty-string `json=""` (lines 857-861), empty choice list, empty grammar string, `json_object=False` — but **no guard for empty dict `json={}`**.

### 2. Falsy-`{}` validation bypass (the key bug)

`backend_xgrammar.py:298` (and identically `sampling_params.py:953`, `backend_outlines.py:175`) uses truthiness instead of `is not None`:

```python
# vllm/v1/structured_output/backend_xgrammar.py:298
if so_params.json:                 # ← if {}: → False! bool({}) == False
    ...
    xgr.Grammar.from_json_schema(schema)   # line 313, NEVER executed for {}
    return
```

`bool({}) == False`, so the entire json validation block is **skipped** for `{}`. `xgr.Grammar.from_json_schema({})` is never called during validation, so the empty schema is never inspected/rejected up front. It falls through to runtime compilation.

The same falsy-`{}` bug appears in 3 places:
- `vllm/v1/structured_output/backend_xgrammar.py:298`
- `vllm/sampling_params.py:953` (auto fallback path)
- `vllm/v1/structured_output/backend_outlines.py:175`

Compare with `vllm/v1/structured_output/request.py:67`, which correctly uses `if params.json is not None:` — the same library mixes truthiness and `is not None`, so `{}` behaves inconsistently between key-construction and validation.

### 3. Empty schema compiles to "any value" grammar → unbounded number

All four backends faithfully compile `{}` per JSON Schema spec (no `type` = accept any value), and the `number` branch is `digit+` with no max-length / no forced terminator:

- xgrammar: `vllm/v1/structured_output/backend_xgrammar.py:81` `compile_json_schema("{}")`
- guidance/llguidance: `vllm/v1/structured_output/backend_guidance.py:230` `grammar_from_json_schema("{}")`
- outlines: `vllm/v1/structured_output/backend_outlines.py:73` `build_regex_from_schema("{}")`
- lm-format-enforcer: same pattern

**Empirical proof with llguidance** (`LLMatcher.try_consume_tokens`, returns count of accepted tokens):

| schema | feed `-`+30 digits | keep feeding digits | digits then `}` | digits then EOS |
|---|---|---|---|---|
| `{}` (empty) | **ok=19 accepted** | **ok=6 always accepted, is_stopped=False** | **ok=0 rejected** (number branch refuses `}` terminator) | ok=1 accepted |
| `{"type":"object"}` | ok=0 rejected (only `{` accepted) | — | — | — |
| `{"type":"string"}` | ok=0 rejected (only `"` accepted) | — | — | — |

Serialized grammar for `{}`:
```json
{"grammars":[{"name":"json_schema","json_schema":{"x-guidance":{"whitespace_flexible":true}},"lark_grammar":null}],"max_tokens":null}
```
Note: no `type` field → "any value". The number branch accepts unlimited digits and **refuses `}` termination** (ok=0). Once the model samples a digit/`-` as the first content token, the FSM enters the number state and the grammar never permits `}`/whitespace to terminate — the model can only keep emitting digits or EOS, and EOS is out-competed by digit logits within `max_tokens`.

### Thinking-phase interaction (why reasoning is normal)

`should_fill_bitmask` (`vllm/v1/structured_output/__init__.py:264-285`) returns `reasoning_ended` (False during thinking) when `enable_in_reasoning=False` (default) — so the bitmask is all-ones during thinking (grammar not enforced, reasoning normal). `should_advance` (line 287-329) returns False during thinking, so `accept_tokens` is never called and the grammar stays in its initial state. When `</think>` is emitted, `reasoning_ended=True`, grammar activates from the **initial state** — and for `{}`, the initial state accepts any-value (including digits), opening the number attractor. The thinking-switch logic itself is correct; the empty schema's initial state is what opens the attractor.

## Why only `json_empty_dict` garbles, other structured outputs are fine

| schema | grammar root | termination |
|---|---|---|
| `{"type":"object","properties":{...}}` | must start `{`, end `}` | structural terminator, `is_terminated()` at `}` |
| `{"type":"string"}` | must start `"`, end `"` | terminator |
| `{"json_object": true}` | constrained to object | normal `{}` |
| `{}` (empty) | any value, **including number** | **number is `digit+` unbounded, no forced terminator** |

Non-empty schemas have a `type` giving the grammar a structural terminator (`}`/`"`/`]`). Empty schema has no `type`, so the number branch is "can continue or end at every step" — once the model leans toward continue, it never terminates.

## Suggested fix

Two independent fixes (either suffices; both recommended):

### Fix A (treat the root cause — recommended): normalize/reject empty schema

In `StructuredOutputsParams.__post_init__` (`vllm/sampling_params.py`):

```python
def __post_init__(self):
    """Validate that some fields are mutually exclusive."""
    # Empty schema {} = accept any JSON value; the number branch is unbounded
    # and cannot terminate, causing a number death-loop. Normalize to
    # json_object (constrain to object, has } terminator). Aligns sglang.
    if isinstance(self.json, dict) and len(self.json) == 0:
        self.json_object = True
        self.json = None
    count = sum([...])
```

### Fix B (remove the falsy-`{}` validation bypass): use `is not None`

In the 3 places that currently use `if so_params.json:`:
- `vllm/v1/structured_output/backend_xgrammar.py:298`
- `vllm/sampling_params.py:953`
- `vllm/v1/structured_output/backend_outlines.py:175`

change to `if so_params.json is not None:` (matching `request.py:67`), so `{}` goes through the normal validation/compilation path instead of being silently skipped.

### Fix A verified

Applied Fix A locally on the image vLLM (`vllm/sampling_params.py`, both prefill and decode pods, restarted vLLM). Same request now returns a normal JSON object:
```json
"content": "{\n  \"name\": \"system_thought\",\n  \"content\": \"收到用户发来的\"你好\"\n}"
```
`finish_reason` is a normal stop (not `length`), `completion_tokens` well under `max_tokens`. reasoning_content unchanged (still normal).

## Additional context

- Original badcase report (Allure): case `test_structured_outputs_json_empty_dict[non_stream/stream]` — both variants garbled, ran 1606s/1291s (full `max_tokens=51200`). The test passed only because its assertion was weak ("should handle or return error" — it accepted either a response or a 400, without checking the response is valid JSON).
- This is a general vLLM guided-decoding pitfall (empty schema = any), not Ascend-specific. Confirmed on GPU by a colleague; sglang does not reproduce (sglang normalizes/rejects empty schema).
- Related vLLM structured-output area: `StructuredOutputsParams` (`vllm/sampling_params.py:40`), `StructuredOutputManager` (`vllm/v1/structured_output/__init__.py`), backends in `vllm/v1/structured_output/backend_*.py`.

### PR DESCRIPTION
**Pull request by fyrsta7**
## Purpose

Fix #52011 by rejecting an empty JSON Schema before structured-output backend selection. The empty schema `{}` is a valid universal JSON Schema, but it leaves number generation unconstrained. If decoding enters the number branch, it may continue producing digits until `max_tokens` and return an unusable response with `finish_reason="length"`.

This intentionally changes empty schemas from an unconstrained request to an HTTP 400 validation error. It covers dictionary input (`json={}`), serialized input (`json="{}"`), and serialized input with surrounding whitespace. Callers should provide a concrete non-empty schema, or use `json_object=True` only when any JSON object matches the intended output.

The Python and Rust frontends now apply the same check before backend compilation or model execution. The Rust request boundary covers raw `structured_outputs`, OpenAI `response_format`, and the gRPC structured-output path. Non-empty schemas continue to pass request validation, and a failed request does not prevent subsequent requests from completing.

This does not duplicate #40099, #51450, #45346, or #47176: those changes respectively address general repetition detection, whitespace-only Rust frontend strings, `json=""`/`json_object=False`, and other Rust validation parity gaps such as null schemas and mutually exclusive constraints. Open PR #51450 touches the same Rust protocol file, but its check only uses `value.trim().is_empty()`; it neither rejects an object `{}` nor parses serialized `"{}"`. A duplicate check on 2026-08-13 found no assignee, claim comment, linked pull request, or open pull request implementing empty-schema rejection.

AI assistance from OpenAI Codex was used to investigate the issue, implement the validation, add tests and documentation, and prepare this description. I reviewed every changed line, validated the behavior end to end, and ran the tests listed below.

## Test Plan

- `.venv/bin/python -m pytest tests/v1/structured_output/test_validation.py -v`
- `.venv/bin/python -m pytest tests/v1/structured_output -m cpu_test -q`
- `.venv/bin/python -m pytest tests/entrypoints/openai/chat_completion/test_chat_completion.py::test_empty_json_schema_returns_bad_request -v -s`
- `cd rust && cargo fmt --all --check`
- `cd rust && cargo test -p vllm-engine-core-client`
- `cd rust && cargo test -p vllm-server`
- `cd rust && cargo clippy -p vllm-engine-core-client -p vllm-server --tests -- -D warnings`
- `git diff --name-only -z origin/main...HEAD | xargs -0 .venv/bin/pre-commit run --files`
- Manual OpenAI API smoke tests against both the Python and Rust frontends with Qwen2.5-1.5B-Instruct, covering dictionary and serialized empty schemas followed by a normal request.
- `git diff --check origin/main...HEAD`

## Test Result

- On unmodified `main`, a direct validation probe accepted `{}`, `"{}"`, and a whitespace-padded `"{}"`.
- Focused validation tests: `9 passed, 14 warnings in 3.91s`.
- Structured-output CPU tests: `25 passed, 23 deselected, 14 warnings in 9.38s`.
- OpenAI API regression test: passed.
- Rust engine-core-client tests: `99 passed`; Rust server tests: `342 passed`; Clippy with warnings denied and `cargo fmt --check` both passed.
- Manual OpenAI API smoke tests with Qwen2.5-1.5B-Instruct produced the same result on both frontends: dictionary `{}` returned HTTP 400, serialized `"{}"` returned HTTP 400, and the immediately following normal request returned HTTP 200.
- All applicable pre-commit hooks passed, including Ruff, markdownlint, typos, mypy for Python 3.10, SPDX, forbidden-import checks, and configuration validation, including Rust formatting.
- `git diff --check origin/main...HEAD` passed.

Model evaluation is not applicable because the affected requests fail during validation before model execution. Successful request generation is not modified. User-facing behavior and the compatibility tradeoff are documented in `docs/features/structured_outputs.md`.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR and the issue it resolves are described above.
- [x] Reproducible unit, integration, lint, and manual test steps are listed.
- [x] Before-and-after validation and HTTP results are listed.
- [x] The necessary structured-output documentation is updated.
</details>

### CODE REVIEW

**Comment by fyrsta7** [comment]
Hi, I rebased this PR onto the latest `main` and resolved the conflict introduced by #51796, preserving both the upstream NUL-regex validation tests and this PR's empty JSON schema tests. The PR is now conflict-free.

Post-rebase validation:
- `tests/v1/structured_output/test_validation.py`: 12 passed
- structured-output CPU suite: 28 passed, 23 deselected
- Rust engine-core-client: 99 passed
- Rust server: 342 passed
- `cargo fmt`, Clippy with warnings denied, changed-files pre-commit, and `git diff --check`: passed

Thank you!

**Comment by fyrsta7** [comment]
Hi, I rebased this PR onto the latest `main` and resolved the conflict introduced by #52394, preserving its structured-output client-error and fallback tests alongside this PR's empty JSON schema validation tests. The concurrent gRPC `skip_special_tokens` change from #52384 is also preserved. The PR is now conflict-free.

Post-rebase validation:
- `tests/v1/structured_output/test_validation.py`: 23 passed
- structured-output CPU suite: 39 passed, 23 deselected
- Rust engine-core-client: 99 passed
- Rust server: 343 passed
- Rust fmt, Clippy with warnings denied, changed-files pre-commit, and `git diff --check`: passed

Thank you!

**Comment by BugenZhao** [comment]
@codex review

**Comment by arpera** [comment]
@fyrsta7, first of all thanks for a PR! I have several questions about your PR:

1. In this PR you decided to reject requests that send empty json schema, but at the same time initial bug report https://github.com/vllm-project/vllm/issues/52011 proposes two different solutions to this problem: (1) to reject such requests; (2) to normalize empty json schema to `{"type": "object"}`. Any motivation why you decided to follow (1) approach and not (2)?

2. I think the problem reported is not limited only to empty json schema but also it might be reproduced by `{"type": "number"}`, `{"type": "integer"}`, `{"type": ["object", "number"]}`, `{"title": "x"}`, `{"description": "any"}`, and so on. So, by rejecting empty json schema we still do not resolve the root cause of the issue and it can be reproduced many other ways. In my opinion the better way would be not to reject such schemas including empty but to set manually `max_tokens` constraint for such a requests to not allow them consume `max_tokens` budget. What do you think about this proposal?
